### ISSUE TITLE
AssertionError (bad cast) from self-type reference

### ISSUE THREAD

**Issue by djspiewak** [lampepfl/dotty#9397]
I don't have anything minimized, but I do have something reproducible. Grab this tag: https://github.com/djspiewak/cats-effect/tree/bug/dotty-0720 and run `sbt coreJVM/test:compile`. You'll see a few errors which look like this:

```
exception while typing this.$asInstanceOf$[
  cats.effect.IO.type & cats.effect.IOCompanionPlatform & 
    (IOCompanionPlatform.this : cats.effect.IOCompanionPlatform)
  
].effectForIO.fromCompletableFuture of class class dotty.tools.dotc.ast.Trees$Select # 586878
exception while typing this.$asInstanceOf$[
  cats.effect.IO.type & cats.effect.IOCompanionPlatform & 
    (IOCompanionPlatform.this : cats.effect.IOCompanionPlatform)
  
].effectForIO.fromCompletableFuture[A] of class class dotty.tools.dotc.ast.Trees$TypeApply # 586885
exception while typing this.$asInstanceOf$[
  cats.effect.IO.type & cats.effect.IOCompanionPlatform & 
    (IOCompanionPlatform.this : cats.effect.IOCompanionPlatform)
  
].effectForIO.fromCompletableFuture[A](fut) of class class dotty.tools.dotc.ast.Trees$Apply # 586886
exception while typing def fromCompletableFuture[A](
  fut: cats.effect.IO[java.util.concurrent.CompletableFuture[A]]
): cats.effect.IO[A] = 
  this.$asInstanceOf$[
    cats.effect.IO.type & cats.effect.IOCompanionPlatform & 
      (IOCompanionPlatform.this : cats.effect.IOCompanionPlatform)
    
  ].effectForIO.fromCompletableFuture[A](fut) of class class dotty.tools.dotc.ast.Trees$DefDef # 586887
exception while typing @scala.annotation.internal.SourceFile(
  
    "/Users/daniel/Development/Scala/cats-effect/series-3.x/core/jvm/src/main/scala/cats/effect/IOCompanionPlatform.scala"
```

The exception stack trace:

```
[error] java.lang.AssertionError: assertion failed: bad cast: (this.asInstanceOf[cats.effect.IO$]:cats.effect.IO$).effectForIO().asInstanceOf[(): cats.effect.kernel.Effect]
[error] dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
[error] dotty.tools.dotc.ast.tpd$TreeOps$.asInstance$extension(tpd.scala:958)
[error] dotty.tools.dotc.transform.Erasure$Boxing$.op$3(Erasure.scala:334)
[error] dotty.tools.dotc.transform.Erasure$Boxing$.cast(Erasure.scala:336)
[error] dotty.tools.dotc.transform.Erasure$Typer.recur$1(Erasure.scala:705)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedSelect(Erasure.scala:709)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2362)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2451)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:124)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2518)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2527)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2530)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2641)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedTypeApply(Erasure.scala:724)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2410)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2452)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:124)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2518)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2527)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2530)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2641)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedApply(Erasure.scala:746)
[[[[snip]]]]
```

**Comment by djspiewak** [comment]
Oh here's a fun one! So the origin of this is the following class:

```scala
abstract private[effect] class IOCompanionPlatform { self: IO.type =>

  def fromCompletableFuture[A](fut: IO[CompletableFuture[A]]): IO[A] =
    effectForIO.fromCompletableFuture(fut)
}
```

If I add an explicit ascription, I can get Dotty to spin forever:

```scala
abstract private[effect] class IOCompanionPlatform { self: IO.type =>

  def fromCompletableFuture[A](fut: IO[CompletableFuture[A]]): IO[A] =
    (effectForIO: Effect[IO]).fromCompletableFuture(fut)
}
```

I had to `kill -9` sbt.

**Comment by djspiewak** [comment]
Sorry for the noise. This is fixed in `0.26.0-bin-20200717-86cc7f8-NIGHTLY`
